pkg/controller/windowsmachineconfig/windowsmachineconfig_controller.go

```diff
@@ -398,3 +416,97 @@ func (e *reconcileError) reason() string {
+func retrievePrivateKey() (*rsa.PrivateKey, error) {
+	privateKeyBytes, err := ioutil.ReadFile(wkl.PrivateKeyPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find private key from path: %v with: %v", wkl.PrivateKeyPath, err)
+	}
+
+	privateKeyBlock, _ := pem.Decode(privateKeyBytes)
+	if privateKeyBlock == nil {
+		return nil, fmt.Errorf("failed to decode retrieved private key: %v", wkl.PrivateKeyPath)
+	}
+
+	var parsedKey interface{}
+	if parsedKey, err = x509.ParsePKCS1PrivateKey(privateKeyBlock.Bytes); err != nil {
+		return nil, fmt.Errorf("failed to parse private key with: %v", err)
+	}
+
+	privateKey, ok := parsedKey.(*rsa.PrivateKey)
+	if !ok {
+		return nil, fmt.Errorf("failed to cast private key as rsa key type: %v", err)
+	}
+	return privateKey, nil
+}
+
+// createUserDataSecret creates a secret in 'openshift-machine-api' namespace
+// This secret will be used to inject cloud provider user data for creating
+// windows machines
+func (r *ReconcileWindowsMachineConfig) createUserDataSecret() error {
+
+	userDataSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "windows-user-data",
+			Namespace: "openshift-machine-api",
+		},
+		Data: map[string][]byte{
+			"userData": []byte(`<powershell>
+	$url = "https://raw.githubusercontent.com/ansible/ansible/devel/examples/scripts/ConfigureRemotingForAnsible.ps1"
+	$file = "$env:temp\ConfigureRemotingForAnsible.ps1"
+	(New-Object -TypeName System.Net.WebClient).DownloadFile($url,  $file)
+	& $file
+	Add-WindowsCapability -Online -Name OpenSSH.Server~~~~0.0.1.0
+	New-NetFirewallRule -DisplayName "` + types.FirewallRuleName + `"
+	-Direction Inbound -Action Allow -Protocol TCP -LocalPort ` + types.ContainerLogsPort + ` -EdgeTraversalPolicy Allow
+	Install-PackageProvider -Name NuGet -MinimumVersion 2.8.5.201 -Force
+	Install-Module -Force OpenSSHUtils -Scope AllUsers
+	Set-Service -Name ssh-agent -StartupType ‘Automatic’
+	Set-Service -Name sshd -StartupType ‘Automatic’
+	Start-Service ssh-agent
+	Start-Service sshd
+	$pubKeyConf = (Get-Content -path C:\ProgramData\ssh\sshd_config) -replace '#PubkeyAuthentication yes','PubkeyAuthentication yes'
+	$pubKeyConf | Set-Content -Path C:\ProgramData\ssh\sshd_config
+ 	$passwordConf = (Get-Content -path C:\ProgramData\ssh\sshd_config) -replace '#PasswordAuthentication yes','PasswordAuthentication yes'
+	$passwordConf | Set-Content -Path C:\ProgramData\ssh\sshd_config
+	$authFileConf = (Get-Content -path C:\ProgramData\ssh\sshd_config) -replace 'AuthorizedKeysFile __PROGRAMDATA__/ssh/administrators_authorized_keys','#AuthorizedKeysFile __PROGRAMDATA__/ssh/administrators_authorized_keys'
+	$authFileConf | Set-Content -Path C:\ProgramData\ssh\sshd_config
+	$pubKeyLocationConf = (Get-Content -path C:\ProgramData\ssh\sshd_config) -replace 'Match Group administrators','#Match Group administrators'
+	$pubKeyLocationConf | Set-Content -Path C:\ProgramData\ssh\sshd_config
+	Restart-Service sshd
```

Comment:
Can't you set the configuration and just start the service here? I don't see any reason to start the service and restart it again here.